cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedTable.java

```diff
@@ -518,36 +558,49 @@ public void write(byte[] bytes, Put put) {
     put(put);
   }
 
-  private class IndexScanner implements Scanner {
+  private abstract class AbstractIndexScanner implements Scanner {
     // scanner over index table
     private final Scanner baseScanner;
-    private final byte[] rowKeyPrefix;
+    private final byte[] column;
 
-    public IndexScanner(Scanner baseScanner, byte[] rowKeyPrefix) {
+    public AbstractIndexScanner(Scanner baseScanner, byte[] column) {
       this.baseScanner = baseScanner;
-      this.rowKeyPrefix = rowKeyPrefix;
+      this.column = column;
     }
 
+    // checks if a particular column value matches a criteria defined by the implementing class
+    // return false indicates to skip the corresponding row
+    protected abstract boolean matches(byte[] columnValue);
+
     @Nullable
     @Override
     public Row next() {
       // TODO: retrieve results in batches to minimize RPC overhead (requires multi-get support in table)
-      Row dataRow = null;
       // keep going until we hit a non-null, non-empty data row, or we exhaust the index
-      while (dataRow == null) {
-        Row indexRow = baseScanner.next();
-        if (indexRow == null) {
-          // end of index
-          return null;
-        }
+      Row indexRow;
+      while ((indexRow = baseScanner.next()) != null) {
         byte[] rowkey = indexRow.get(IDX_COL);
-        // verify that datarow matches the expected row key to avoid issues with column name or value
-        // containing the delimiter used
-        if (rowkey != null && Bytes.equals(indexRow.getRow(), Bytes.add(rowKeyPrefix, rowkey))) {
-          dataRow = table.get(rowkey);
+        if (rowkey == null) {
+          continue;
```

Comment:
what does it mean if the rowkey is null? Can that ever happen? I feel like that indicates data corruption. Perhaps issue a warning in this case?